File: crates/fornax-py/fornax/_dcraw.py
from enum import Enum
from pathlib import Path
from typing import Self

from pydantic import BaseModel, FilePath, PositiveInt, StrictBool, field_serializer


class DCRawHighlightMode(int, Enum):
    """Highlight modes."""

    Clip = 0
    Ignore = 1
    Blend = 2
    Reconstruct3 = 3
    Reconstruct4 = 4
    Reconstruct5 = 5
    Reconstruct6 = 6
    Reconstruct7 = 7
    Reconstruct8 = 8
    Reconstruct9 = 9


class DCRawUseCameraMatrix(int, Enum):
    NotUse = 0
    EmbeddedProfile = 1
    EmbeddedData = 3


class DCRawOutputBps(int, Enum):
    _8bit = 8
    _16bit = 6


class DCRawOutputColor(int, Enum):
    """Color spaces."""

    RAW = 0
    SRGB = 1
    ADOBE = 2
    WIDE = 3
    PROPHOTO = 4
    XYZ = 5
    ACES = 6
    P3D65 = 7
    REC2020 = 8


class DCRawOutputTiff(int, Enum):
    _None = -1
    Ppm = 0
    Tiff = 1


class DCRawUserFlip(int, Enum):
    _None = 0
    Rotate180 = 3
    CCW90 = 5
    CW90 = 6


class DCRawUserQual(int, Enum):
    """Identifiers for demosaic algorithms."""

    LINEAR = 0
    VNG = 1
    PPG = 2
    AHD = 3
    DCB = 4
    # comment GPL algorithm
    # MODIFIED_AHD = 5
    # AFD = 6
    # VCD = 7
    # VCD_MODIFIED_AHD = 8
    # LMMSE = 9
    # AMAZE = 10
    DHT = 11
    AAHD = 12


class DCRawUseFujiRotate(int, Enum):
    UseRotate = -1
    NotUse = 0


class DCRawFbddNoiserd(int, Enum):
    """FBDD noise reduction modes."""

    OFF = 0
    LIGHT = 1
    FULL = 2


class DCRawParams(BaseModel):
    """
    Create `rawpy.Params`.

    References
    ----------
    - `Libraw libraw_output_params_t <https://www.libraw.org/docs/API-datastruct-eng.html#libraw_output_params_t>`_

    """

    greybox: tuple[PositiveInt, PositiveInt, PositiveInt, PositiveInt] | None = None
    cropbox: tuple[PositiveInt, PositiveInt, PositiveInt, PositiveInt] | None = None
    aber: tuple[float, float] | None = None
    gamm: tuple[float, float] | None = None
    user_mul: tuple[float, float, float, float] | None = None
    bright: float | None = None
    threshold: float | None = None
    half_size: StrictBool | None = None
    four_color_rgb: StrictBool | None = None
    highlight: DCRawHighlightMode | None = None
    use_auto_wb: StrictBool | None = None
    use_camera_wb: StrictBool | None = None
    use_camera_matrix: DCRawUseCameraMatrix | None = None
    output_color: DCRawOutputColor | None = None
    output_profile: FilePath | None = None
    camera_profile: FilePath | None = None
    bad_pixels: FilePath | None = None
    dark_frame: FilePath | None = None
    output_bps: DCRawOutputBps | None = None
    output_tiff: DCRawOutputTiff | None = None
    user_flip: DCRawUserFlip | None = None
    user_qual: DCRawUserQual | None = None
    user_black: int | None = None
    user_cblack: tuple[int, int, int, int] | None = None
    user_sat: int | None = None
    med_passes: int | None = None
    no_auto_bright: StrictBool | None = None
    auto_bright_thr: float | None = None
    adjust_maximum_thr: float | None = None
    use_fuji_rotate: DCRawUseFujiRotate | None = None
    green_matching: StrictBool | None = None
    dcb_iterations: int | None = None
    dcb_enhance_fl: int | None = None
    fbdd_noiserd: DCRawFbddNoiserd | None = None
    exp_correc: int | None = None
    exp_shift: float | None = None
    exp_preser: StrictBool | None = None
    use_rawspeed: StrictBool | None = None
    no_auto_scale: StrictBool | None = None
    no_interpolation: StrictBool | None = None

    @field_serializer(
        "highlight",
        "use_camera_matrix",
        "output_color",
        "output_bps",
        "output_tiff",
        "user_flip",
        "user_qual",
        "use_fuji_rotate",
        "fbdd_noiserd",
        when_used="json",
    )
    def _serialize_special_field(self, v: Enum):
        return v.value

    @classmethod
    def read_json(cls, json_file: str | Path) -> Self:
        """
        Create a model from json file.

        Parameters
        ----------
        json_file
            Path of json file.

        Returns
        -------
        DCRawParams
        """
        json_file = Path(json_file)
        assert json_file.exists()
        assert json_file.is_file()
        assert json_file.suffix.endswith("json")
        return cls.model_validate_json(json_data=json_file.read_text("utf-8"))

    def write_json(self, json_file: str | Path):
        """
        Dump a model to json file.

        Parameters
        ----------
        json_file
            Path of json file.
        """
        json_file = Path(json_file)
        assert json_file.suffix.endswith("json")
        content = self.model_dump_json(indent=4)
        json_file.write_text(content, "utf-8")

File: crates/libraw/src/extension/dcraw.rs
use std::sync::Arc;

use envoy::ToCString;

use crate::{DCRawOutputBps, DCRawOutputColor, DCRawParams, ImgdataPointer, LibrawError};

impl DCRawParams {
    #[allow(unused_mut)]
    pub(crate) fn set_output_params(
        &self,
        mut imgdata: Arc<ImgdataPointer>,
    ) -> Result<(), LibrawError> {
        self.greybox
            .inspect(|v| unsafe { (**imgdata).params.greybox = *v });
        self.cropbox
            .inspect(|v| unsafe { (**imgdata).params.cropbox = *v });
        self.aber.inspect(|v| unsafe {
            (**imgdata).params.aber[0] = v[0];
            (**imgdata).params.aber[2] = v[1];
        });
        self.gamm.inspect(|v| unsafe {
            (**imgdata).params.gamm[0] = v[0];
            (**imgdata).params.gamm[1] = v[1];
        });
        self.user_mul
            .inspect(|v| unsafe { (**imgdata).params.user_mul = *v });
        self.bright
            .inspect(|v| unsafe { (**imgdata).params.bright = *v });
        self.threshold
            .inspect(|v| unsafe { (**imgdata).params.threshold = *v });
        self.half_size
            .inspect(|v| unsafe { (**imgdata).params.half_size = *v as i32 });
        self.four_color_rgb
            .inspect(|v| unsafe { (**imgdata).params.four_color_rgb = *v as i32 });
        self.highlight
            .inspect(|v| unsafe { (**imgdata).params.highlight = (*v) as i32 });
        self.use_auto_wb
            .inspect(|v| unsafe { (**imgdata).params.use_auto_wb = *v as i32 });
        self.use_camera_wb
            .inspect(|v| unsafe { (**imgdata).params.use_camera_wb = *v as i32 });
        self.use_camera_matrix
            .inspect(|v| unsafe { (**imgdata).params.use_camera_matrix = (*v) as i32 });
        self.output_color
            .inspect(|v| unsafe { (**imgdata).params.output_color = (*v) as i32 });

        if let Some(v) = self.output_profile.as_ref() {
            unsafe {
                (**imgdata).params.output_profile = v.to_str().unwrap().to_cstring()?.into_raw();
            }
        }
        if let Some(v) = self.camera_profile.as_ref() {
            unsafe {
                (**imgdata).params.camera_profile = v.to_str().unwrap().to_cstring()?.into_raw();
            }
        }

        if let Some(v) = self.bad_pixels.as_ref() {
            unsafe {
                (**imgdata).params.bad_pixels = v.to_str().unwrap().to_cstring()?.into_raw();
            }
        }

        if let Some(v) = self.dark_frame.as_ref() {
            unsafe {
                (**imgdata).params.dark_frame = v.to_str().unwrap().to_cstring()?.into_raw();
            }
        }

        self.output_bps
            .inspect(|v| unsafe { (**imgdata).params.output_bps = (*v) as i32 });
        self.output_tiff
            .inspect(|v| unsafe { (**imgdata).params.output_tiff = (*v) as i32 });
        self.user_flip
            .inspect(|v| unsafe { (**imgdata).params.user_flip = (*v) as i32 });
        self.user_qual
            .inspect(|v| unsafe { (**imgdata).params.user_qual = (*v) as i32 });
        self.user_black
            .inspect(|v| unsafe { (**imgdata).params.user_black = *v });
        self.user_cblack
            .inspect(|v| unsafe { (**imgdata).params.user_cblack = *v });
        self.user_sat
            .inspect(|v| unsafe { (**imgdata).params.user_sat = *v });
        self.med_passes
            .inspect(|v| unsafe { (**imgdata).params.med_passes = *v });
        self.no_auto_bright
            .inspect(|v| unsafe { (**imgdata).params.no_auto_bright = *v as i32 });
        self.auto_bright_thr
            .inspect(|v| unsafe { (**imgdata).params.auto_bright_thr = *v });
        self.adjust_maximum_thr
            .inspect(|v| unsafe { (**imgdata).params.adjust_maximum_thr = *v });
        self.use_fuji_rotate
            .inspect(|v| unsafe { (**imgdata).params.use_fuji_rotate = *v as i32 });
        self.dcb_iterations
            .inspect(|v| unsafe { (**imgdata).params.dcb_iterations = *v });
        self.dcb_enhance_fl
            .inspect(|v| unsafe { (**imgdata).params.dcb_enhance_fl = *v });
        self.fbdd_noiserd
            .inspect(|v| unsafe { (**imgdata).params.fbdd_noiserd = (*v) as i32 });
        self.exp_correc
            .inspect(|v| unsafe { (**imgdata).params.exp_correc = *v });
        self.exp_shift
            .inspect(|v| unsafe { (**imgdata).params.exp_shift = *v });
        self.exp_preser
            .inspect(|v| unsafe { (**imgdata).params.exp_preser = *v });
        // self.use_rawspeed
        //     .inspect(|v| unsafe { (**imgdata).params.use_rawspeed = *v });
        // self.use_dng_sdk
        //     .inspect(|v| unsafe { (**imgdata).params.use_dng_sdk = *v });
        self.no_auto_scale
            .inspect(|v| unsafe { (**imgdata).params.no_auto_scale = *v as i32 });
        self.no_interpolation
            .inspect(|v| unsafe { (**imgdata).params.no_interpolation = *v as i32 });
        Ok(())
    }
}
// presets
impl DCRawParams {
    /// Match output to cg workflow.
    /// - `gamm` = `[1.0, 1.0]`
    /// - `output_color`: ACES
    /// - `output_bps`: 16bit
    pub fn preset_cg() -> Self {
        Self {
            gamm: Some([1.0, 1.0]),
            output_color: Some(DCRawOutputColor::ACES),
            output_bps: Some(DCRawOutputBps::_16bit),
            ..Default::default()
        }
    }
}
